### DISCUSSION TITLE
Current & Upcoming Versions: Which should we use?

### DISCUSSION THREAD

**Discussion by ahoebeke** [directus/directus#3043] [Archive]
For someone new to Directus, reading about all possible versions gets a bit confusing.

1. the current version directus/directus
2. directus/next (upcoming NodeJS version) 
3. directus/api-next (upcoming Laravel/PHP based version)

I would love to use option 2., to benefit from Postgresql support. I also like the idea that it's node/vuejs/typescript based, seems like a great fit for this sort of project.

Can there be a clarification or official comparison and clarification of the various versions and which ones to use in what scenario?

**Comment by danieljwestman** [comment]
+1 on this 👍🏻

Would be nice with some explanations about this releases and the roadmap, even if it's just some rough estimations 😊

In the best of worlds I would like to run the latest version of the App (v9 node vue) but an PHP version of the API (in that way it would be possible to create a multisite/multitenancy structure with different .env-vars depending on host/slug.

**Comment by benhaynes** [comment]
100%! I'll provide a quick update on all of this here, but it will all be crystal clear within the next week, when we update/consolidate all the repos.

## Directus 8
Written in PHP+Zend. Only supports MySQL and MariaDB. Hasn't been _actively_ worked on in about a year except small fixes. This is the current production version, and is spread across several repos: `directus/directus`, `directus/app`, `directus/api`, `directus/title-formatter`, `directus/docs`, etc. It will soon be replaced by...

## Directus 9
Written in Node.js+Vue.js. Supports MySQL, PostgreSQL, SQLite, MS-SQL Server, OracleDB, and more. Our entire team has been working full-time on getting this ready. It has _huge_ improvements in terms of speed, features, stability, and more. It is currently on the `directus/next` repo — and when we shift to RC.1, we will be moving this "monorepo" to a branch on `directus/directus`, and consolidating ALL the above repos.

Yes, there is a BIG shift moving from PHP to node. We'll be discussing this decision more in upcoming articles. There are good reasons. But to be clear, v8 will still be available, and we're also working on...

## Directus 9 (PHP)
Written in PHP+Laravel (complete refactor). In terms of features and support, it will be the same as v9 Node. Node will be the primary software for the organization, and will define the Directus specification, but we didn't want to abandon our PHP community that has been with us for so long. It is currently on `directus/api-next`... this naming was due to it _only_ being the API (it will re-use the same Vue.js App within the node repo). This may stay its own repo (with a clearer name), or become a branch in `directus/directus`. Since this codebase must follow the node port spec, it's a few months behind the official v9 release.

This total refactor was required because of technical debt in the v8 codebase that piled up over the past decade, and due to our database abstraction layer (ZendDB) being deprecated. To be clear, we are a small team and can not (at our current size) manage both these codebases. So our goal as of now, is to build the PHP-Laravel option, and then have it be community-driven from that point on. If there is enough interest/usage, we'll reassess adding additional resources here... but as of now our focus will be the node.js port.

## Ecosystem
We also have several other long-awaited things coming down the pipe for this v9 release:

* New Marketing Site (it's beautiful)
* New Directus Cloud Enterprise (already up and running)
* New Directus Cloud On-Demand (still in work)
* New v9 Docs (in node monorepo... still in work)
* New video tutorial series on YouTube (still in work)
* New positioning (marketing only, from Headless CMS to Data Platform)
* New Community Chat (from Slack to [Discord](https://discord.gg/directus))
* Bigger Team (been growing over the past few months)
* New Docker Image and many other install options

---

We know this is a big and confusing change, but consider all this to be in stealth mode for the next few weeks... and then it will all be cleared up when v9 is out. The only reason we're starting to push things and announce Betas/RCs is so we can get more user testing.

Thanks for bearing with us... we are _so excited_ to share our efforts with you!! ❤️ 

— The Directus Team ✨🐰✨

**Comment by danieljwestman** [comment]
Thanks - great answer and response!

Will the cloud storage option/credentials (S3) be in the PHP API part or in the App Node part?

With this we could easy be building custom multitenancy/multisite in v9! (From what I can understand)

@benhaynes

**Reply by rijkvanzanten** [reply]
> Will the cloud storage option/credentials (S3) be in the PHP API part or in the App Node part?

The credentials are server side, regardless of using v8, v9 (php or node)

**Reply by danieljwestman** [reply]
Hm, well in v9 (Node App/API) you put the credentials for db and storage in the .env file which is loaded when starting up Node. You aren’t able to change this on runtime. Have I missed something? @rijkvanzanten

**Reply by rijkvanzanten** [reply]
That's correct. You set it up, and then (re)start the API. This does not mean however that these env vars are visible from the _app_ as that would leak it to the client side.

**Reply by danieljwestman** [reply]
Yes, I do understand that. But with the ”new” possibility to run the app on node and the api on php we could put some dynamic (during the node runtime) credential handling on server-side. Right?

**Reply by rijkvanzanten** [reply]
I think you misunderstood that setup. The app itself is a static SPA, so it'll be the Node API that's "swapped out" for the PHP version once that's ready.

**Reply by danieljwestman** [reply]
No, I think I understand the set-up 😁

We will have the possibility to run the App on Node and the API och PHP. The app will be static and call the "remote" API. In the (PHP) API we will be able to to something with each request. For an example check for cookies and based on that use different credentials for db and storage. Am I right? 😁

**Reply by rijkvanzanten** [reply]
> We will have the possibility to run the App on Node

The app itself is a static html dist, so you can technically do with that as you please. That being said, the app expects to be served from it's own API under `/admin`, so your idea of having the API remote hosted separately from the app won't work.

**Reply by danieljwestman** [reply]
I don't need the API to be remote (the DB will be, but that has nothing to do with this), it was a sloppy formulation from my side. But it will be "decoupled" server-side PHP scripts that runs the API, and "in the (PHP) API we will be able to do something with each request from the app. For an example check for cookies and based on that use different credentials for db and storage"

Correct?

**Reply by rijkvanzanten** [reply]
I'm not following what you're trying to achieve haha...

In both APIs (Node/PHP) you're able to add Hooks that allow you to do any custom logic on every/any request. That being said, both APIs won't support dynamically swapping databases in runtime.
